### PR TITLE
feat: uplink fleet simulation

### PR DESCRIPTION
**Pull request by de-sh**
Closes #

### Changes
Script to start multiple uplink instances, each with an app connecting to appropriate tcp port and sending diverse set of randomly generated data at expected frequencies using faker and delayqueue. Pushes data in sequence `1..RESET_LIMIT` and then restarts from 1.

### Why?
For use in verifying pipeline stability.

### Trials Performed
Trialed against stage, simulator generates random values for BMS, IMU, Device Shadow, etc. All expected streams are being pushed to appropriate clickhouse tables with appropriate device ids, confirming expected behaviour. The streams are pushing at expected frequency with appropriate values and sequences are being reset on reaching the appropriate limit:
```
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"imu\",\"sequence\":1499,\"timestamp\":1676892368036,\"ax\":2.7443790400843078,\"ay\":2.5413491510178874,\"az\":9.81757102330543,\"magx\":-34.19967377632463,\"magy\":-35.138039600961946,\"magz\":-22.5764604930371,\"pitch\":0.8480097839729313,\"roll\":0.9851863196737695,\"yaw\":0.8733544085109946}"
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"imu\",\"sequence\":1500,\"timestamp\":1676892368138,\"ax\":2.642631341284713,\"ay\":2.7953299816758976,\"az\":9.814011607353033,\"magx\":-39.671783701181646,\"magy\":-42.015133368904955,\"magz\":-27.3603748385814,\"pitch\":0.8207438001881766,\"roll\":0.9077028562654432,\"yaw\":0.8607003651391968}"
16:56:08 [TRACE] (2) uplink::base::bridge::stream: Flushing stream name: imu, topic: /tenants/demo/devices/10/events/imu/jsonarray
16:56:08 [TRACE] (3) uplink::base::serializer::serializer: Data received on stream: imu; message count = 100; batching latency = 0
16:56:08 [DEBUG] (3) uplink::base::serializer::serializer: publishing on /tenants/demo/devices/10/events/imu/jsonarray with size = 26815
16:56:08 [DEBUG] (4) uplink::base::mqtt: Outgoing = Publish(19)
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"imu\",\"sequence\":1,\"timestamp\":1676892368239,\"ax\":2.0806714505122734,\"ay\":1.984093216300864,\"az\":9.813324961017106,\"magx\":-33.35847532250173,\"magy\":-25.190440889182184,\"magz\":-32.22056668802537,\"pitch\":0.9771414389875892,\"roll\":0.9809648930528598,\"yaw\":0.8320593394524081}"
16:56:08 [DEBUG] (2) uplink::base::bridge::stream: Sequence number anomaly! [1, 1500
16:56:08 [TRACE] (2) uplink::collector::utils::streams: Initialized stream buffer for imu
16:56:08 [DEBUG] (4) uplink::base::mqtt: Incoming = PubAck(PubAck { pkid: 19 })
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"bms\",\"sequence\":3591,\"timestamp\":1676892368286,\"ambient_temperature\":35.75934313372659,\"cell_balancing_status\":1,\"cell_temp_1\":42.436028652374986,\"cell_temp_2\":40.12319810140357,\"cell_temp_3\":41.990672387444114,\"cell_temp_4\":40.40003195403621,\"cell_temp_5\":40.61499816117445,\"cell_temp_6\":41.15523428198003,\"cell_temp_7\":42.8154804434838,\"cell_temp_8\":41.168530259516956,\"cell_thermistor_count\":8,\"cell_voltage_1\":3.132262904288695,\"cell_voltage_10\":3.0695124619413052,\"cell_voltage_11\":3.0119465925863906,\"cell_voltage_12\":3.1502368745903895,\"cell_voltage_13\":3.1136980212154235,\"cell_voltage_14\":3.055817795901334,\"cell_voltage_15\":3.0901667226918623,\"cell_voltage_16\":3.1879888050090943,\"cell_voltage_2\":3.0582104381163657,\"cell_voltage_3\":3.1215319903942467,\"cell_voltage_4\":3.124646696733153,\"cell_voltage_5\":3.0354536062076254,\"cell_voltage_6\":3.193899677422407,\"cell_voltage_7\":3.076441152665408,\"cell_voltage_8\":3.0024039947818197,\"cell_voltage_9\":3.1937669148651184,\"cell_voltage_count\":16,\"mosfet_status\":1,\"mosfet_temperature\":44.966991310863605,\"pack_available_energy\":2359,\"pack_consumed_energy\":2101,\"pack_current\":19.341898863291796,\"pack_cycle_count\":106,\"pack_fault\":0,\"pack_soc\":87.73387986872777,\"pack_soh\":9.863119365053105,\"pack_sop\":9.576427659077662,\"pack_status\":1,\"pack_voltage\":95.43105474788769,\"periodicity_ms\":250}"
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"motor\",\"sequence\":3591,\"timestamp\":1676892368286,\"motor_current\":23.954423416157805,\"motor_rpm\":5383,\"motor_temperature1\":40.95468244884805,\"motor_temperature2\":42.861829515213344,\"motor_temperature3\":41.08962125800887,\"motor_voltage\":95.1039506811974}"
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"imu\",\"sequence\":2,\"timestamp\":1676892368341,\"ax\":2.7673686376683033,\"ay\":1.030709075879123,\"az\":9.79342869244284,\"magx\":-16.73312715215157,\"magy\":-24.90278046038084,\"magz\":-15.31221618499599,\"pitch\":0.8061066684651595,\"roll\":0.9787696020473641,\"yaw\":0.9666228039745777}"
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"imu\",\"sequence\":3,\"timestamp\":1676892368442,\"ax\":2.416684124436374,\"ay\":1.0608185042620142,\"az\":9.799454338822034,\"magx\":-43.55140265412396,\"magy\":-17.511977928670607,\"magz\":-33.48469991045331,\"pitch\":0.9117577629777489,\"roll\":0.8039555242095381,\"yaw\":0.9921938199958106}"
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"bms\",\"sequence\":3592,\"timestamp\":1676892368537,\"ambient_temperature\":35.318554631798506,\"cell_balancing_status\":1,\"cell_temp_1\":40.592470391997125,\"cell_temp_2\":40.09863146975201,\"cell_temp_3\":41.81141405938229,\"cell_temp_4\":41.012290195682176,\"cell_temp_5\":40.054725348197735,\"cell_temp_6\":42.12877350387498,\"cell_temp_7\":42.72945534183206,\"cell_temp_8\":42.5259784744151,\"cell_thermistor_count\":8,\"cell_voltage_1\":3.1369423693547818,\"cell_voltage_10\":3.1513365493183096,\"cell_voltage_11\":3.0008729082434615,\"cell_voltage_12\":3.0482998247827156,\"cell_voltage_13\":3.0572614128195914,\"cell_voltage_14\":3.06702554889898,\"cell_voltage_15\":3.1561913500844803,\"cell_voltage_16\":3.0378823914885102,\"cell_voltage_2\":3.107853168591729,\"cell_voltage_3\":3.187621368514448,\"cell_voltage_4\":3.1444900625404157,\"cell_voltage_5\":3.1050718220004443,\"cell_voltage_6\":3.162333005123527,\"cell_voltage_7\":3.087781370364208,\"cell_voltage_8\":3.0410771274044563,\"cell_voltage_9\":3.05048191418622,\"cell_voltage_count\":16,\"mosfet_status\":1,\"mosfet_temperature\":43.77367351988115,\"pack_available_energy\":2105,\"pack_consumed_energy\":2820,\"pack_current\":18.107684556204084,\"pack_cycle_count\":147,\"pack_fault\":0,\"pack_soc\":82.34891575971564,\"pack_soh\":9.725005780933857,\"pack_sop\":9.753726774424917,\"pack_status\":1,\"pack_voltage\":95.61246890047764,\"periodicity_ms\":250}"
16:56:08 [TRACE] (1) uplink::collector::tcpjson: 1: Received line = "{\"stream\":\"motor\",\"sequence\":3592,\"timestamp\":1676892368538,\"motor_current\":21.793679836988943,\"motor_rpm\":8785,\"motor_temperature1\":44.57014662417383,\"motor_temperature2\":43.188467367574304,\"motor_temperature3\":41.528914651063886,\"motor_voltage\":95.36221805148776}"
```

#### Running actions on fleet
A `send_file` action triggered from platform for connected devices with id 1, 5, 10 was confirmed to have successfully completed by the following uplink log entries(similar sequence of events observed in all 3 devices, with only device 1 failing initially as it had been down during action init. No other fleet uplinks, i.e. 2,3,4,6, etc received or sent responses for the above action) and associated visualization on the platform:
```
13:54:32 [0m[34m[INFO ] [0m(4) uplink::base::mqtt: Action = Action { device_id: None, action_id: "80", kind: "process", name: "send_file", payload: "{\"file_name\":\"Screenshot from 2023-02-21 09-42-20.png\",\"id\":\"65e1872b-ac88-4cb0-8ddd-d64ae8b4f30b\",\"url\":\"https://firmware.stage.bytebeam.io/api/v1/file/65e1872b-ac88-4cb0-8ddd-d64ae8b4f30b/artifact\"}" }
13:54:32 [0m[34m[INFO ] [0m(2) uplink::base::bridge::bridge: Received action: "80"
13:54:32 [0m[34m[INFO ] [0m(2) uplink::base::bridge::bridge: Action response = ActionResponse { action_id: "80", device_id: None, sequence: 0, timestamp: 1676967872792, state: "Received", progress: 0, errors: [], done_response: None }
13:54:32 [0m[36m[DEBUG] [0m(2) uplink::base::bridge::stream: Sequence number anomaly! [0, 0
13:54:32 [0m[34m[INFO ] [0m(2) uplink::base::bridge::bridge: Action response = ActionResponse { action_id: "80", device_id: None, sequence: 1, timestamp: 1676967872792, state: "Downloading", progress: 0, errors: [], done_response: None }
13:54:32 [0m[36m[DEBUG] [0m(3) uplink::base::serializer::serializer: publishing on /tenants/demo/devices/1/action/status with size = 103
13:54:32 [0m[36m[DEBUG] [0m(3) uplink::base::serializer::serializer: publishing on /tenants/demo/devices/1/action/status with size = 106
13:54:32 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Outgoing = Publish(11)
13:54:32 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Outgoing = Publish(12)
13:54:32 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Incoming = PubAck(PubAck { pkid: 11 })
13:54:32 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Incoming = PubAck(PubAck { pkid: 12 })
13:54:33 [0m[34m[INFO ] [0m(7) uplink::collector::downloader: Downloading from https://firmware.stage.bytebeam.io/api/v1/file/65e1872b-ac88-4cb0-8ddd-d64ae8b4f30b/artifact into /var/tmp/ota/1/send_file/Screenshot from 2023-02-21 09-42-20.png/artifact
13:54:33 [0m[34m[INFO ] [0m(7) uplink::collector::downloader: Firmware downloaded successfully
13:54:33 [0m[34m[INFO ] [0m(2) uplink::base::bridge::bridge: Action response = ActionResponse { action_id: "80", device_id: None, sequence: 2, timestamp: 1676967873022, state: "Downloaded", progress: 100, errors: [], done_response: Some(Action { device_id: None, action_id: "80", kind: "process", name: "send_file", payload: "{\"url\":\"https://firmware.stage.bytebeam.io/api/v1/file/65e1872b-ac88-4cb0-8ddd-d64ae8b4f30b/artifact\",\"version\":\"Screenshot from 2023-02-21 09-42-20.png\",\"download_path\":\"/var/tmp/ota/1/send_file/Screenshot from 2023-02-21 09-42-20.png/artifact\"}" }) }
13:54:33 [0m[36m[DEBUG] [0m(3) uplink::base::serializer::serializer: publishing on /tenants/demo/devices/1/action/status with size = 107
13:54:33 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Outgoing = Publish(13)
13:54:33 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Incoming = PubAck(PubAck { pkid: 13 })
...
13:54:42 [0m[34m[INFO ] [0m(2) uplink::base::bridge::bridge: Action response = ActionResponse { action_id: "80", device_id: None, sequence: 0, timestamp: 1676967882096, state: "in_progress", progress: 90, errors: [], done_response: None }
13:54:42 [0m[36m[DEBUG] [0m(2) uplink::base::bridge::stream: Sequence number anomaly! [0, 0
13:54:42 [0m[36m[DEBUG] [0m(3) uplink::base::serializer::serializer: publishing on /tenants/demo/devices/1/action/status with size = 107
13:54:42 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Outgoing = Publish(38)
13:54:42 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Incoming = PubAck(PubAck { pkid: 38 })
13:54:42 [0m[36m[DEBUG] [0m(3) uplink::base::serializer::serializer: publishing on /tenants/demo/devices/1/device_shadow/jsonarray with size = 204
13:54:42 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Outgoing = Publish(39)
13:54:42 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Incoming = PubAck(PubAck { pkid: 39 })
13:54:43 [0m[34m[INFO ] [0m(2) uplink::base::bridge::bridge: Action response = ActionResponse { action_id: "80", device_id: None, sequence: 0, timestamp: 1676967883095, state: "Completed", progress: 100, errors: [], done_response: None }
13:54:43 [0m[36m[DEBUG] [0m(2) uplink::base::bridge::stream: Sequence number anomaly! [0, 0
13:54:43 [0m[36m[DEBUG] [0m(3) uplink::base::serializer::serializer: publishing on /tenants/demo/devices/1/action/status with size = 106
13:54:43 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Outgoing = Publish(40)
13:54:43 [0m[36m[DEBUG] [0m(4) uplink::base::mqtt: Incoming = PubAck(PubAck { pkid: 40 })
```

It was observed that the action went through the stages, in-order: "Received", "Downloading", "Downloaded", "in_progress" and "Completed"

Screenshot from 2023-02-21 14-01-45

### CODE REVIEW

File: simulator/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+    let mut events = DelayQueue::new();
+
+    generate_initial_events(&mut events, Instant::now(), &device);
+
+    loop {
+        select! {
+            line = actions_rx.next() => {
+                let line = line.ok_or(Error::StreamDone)??;
+                let action = serde_json::from_str(&line)?;
+                generate_action_events(action, &mut events);
```
Comment (tekjar) [inline simulator/src/main.rs:604]: This task is blocking the select! loop. Data won't be generated/sent. Spawn this in a task

File: simulator/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+    info!("Successfully sent action response");
+
+    Ok(())
+}
+
+pub async fn process_events(
+    events: &mut DelayQueue<Event>,
+    client: &mut SplitSink<Framed<TcpStream, LinesCodec>, String>,
+) -> Result<(), Error> {
+    if let Some(e) = events.next().await {
```
Comment (tekjar) [inline simulator/src/main.rs:537]: What does `None` signify here? Early return or unwrap if this should never happen

Comment (de-sh) [reply simulator/src/main.rs:537]: `None` in `DelayQueue` means empty schedule, which is possible if there are no events scheduled, highly unlikely, but possible, which is why I decided to keep the way that simulator used to handle it before, i.e. with an `if let`

File: simulator/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+                progress: i * 10,
+                status: String::from("in_progress"),
+                timestamp: now + duration,
+            }),
+            duration,
+        );
+    }
+
+    let duration = Duration::from_secs(10);
+    events.insert(
```
Comment (tekjar) [inline simulator/src/main.rs:573]: Shouldn't we use `insert_at` instead `insert` and check if events are being generated on time?

**Pull request merged**